src/main/kotlin/com/openlattice/indexing/BackgroundExpiredDataDeletionService.kt

```diff
@@ -0,0 +1,251 @@
+        hazelcastInstance: HazelcastInstance,
+        private val indexerConfiguration: IndexerConfiguration,
+        private val elasticsearchApi: ConductorElasticsearchApi,
+        private val auditingManager: AuditingManager,
+        private val dataGraphService: DataGraphService,
+        private val edm: EdmManager
+) {
+    companion object {
+        private val logger = LoggerFactory.getLogger(BackgroundExpiredDataDeletionService::class.java)!!
+    }
+
+    private val propertyTypes: IMap<UUID, PropertyType> = hazelcastInstance.getMap(HazelcastMap.PROPERTY_TYPES.name)
+    private val entitySets: IMap<UUID, EntitySet> = hazelcastInstance.getMap(HazelcastMap.ENTITY_SETS.name)
+    private val expirationLocks: IMap<UUID, Long> = hazelcastInstance.getMap(HazelcastMap.EXPIRATION_LOCKS.name)
+    private var entitySetToPartition: Map<UUID, Set<Int>> = mapOf()
+    private var propertyTypeIdToColumnData: Map<UUID, Pair<IndexType, EdmPrimitiveTypeKind>> = mapOf()
+
+    init {
+        expirationLocks.addIndex(QueryConstants.THIS_ATTRIBUTE_NAME.value(), true)
+    }
+
+    private val taskLock = ReentrantLock()
+
+    @Suppress("UNCHECKED_CAST", "UNUSED")
+    @Scheduled(fixedRate = MAX_DURATION_MILLIS)
+    fun scavengeExpirationLocks() {
+        expirationLocks.removeAll(
+                Predicates.lessThan(
+                        QueryConstants.THIS_ATTRIBUTE_NAME.value(),
+                        System.currentTimeMillis()
+                ) as Predicate<UUID, Long>
+        )
+    }
+
+    @Suppress("UNUSED")
+    @Scheduled(fixedRate = DATA_DELETION_RATE)
+    fun deleteExpiredDataFromEntitySets() {
+        logger.info("Starting background expired data deletion task.")
+        //Keep number of expired data deletion jobs under control
+        if (taskLock.tryLock()) {
+            try {
+                if (indexerConfiguration.backgroundExpiredDataDeletionEnabled) {
+                    val w = Stopwatch.createStarted()
+                    //We shuffle entity sets to make sure we have a chance to work share and index everything
+                    val lockedEntitySets = entitySets.values
+                            .filter { it.hasExpirationPolicy() }
+                            .filter { tryLockEntitySet(it) }
+                            .shuffled()
+
+                    val propertyTypeIds = lockedEntitySets.filter { it.expiration.startDateProperty.isPresent }.map { it.expiration.startDateProperty.get() }.toSet()
+                    if (propertyTypeIds.isNotEmpty()) {
+                        propertyTypeIdToColumnData = propertyTypes.map{it.key to Pair(it.value.postgresIndexType, it.value.datatype)}.toMap()
+                    }
+
+                    val propertiesOfLockedEntitySets = lockedEntitySets
+                            .map { it.id to edm.getPropertyTypesAsMap(edm.getEntityType(it.entityTypeId).properties) }
+                            .toMap()
+
+                    val totalDeleted = lockedEntitySets
```

Comment:
It seems like this could be combined with the above call into one flow. Not 100% sure

Reply:
you are right! combined with getting of property types. I left the generation of the locked entity sets separate tho so `lockedEntitySets.forEach(this::deleteIndexingLock)` can still be used without having to filter through all of the entity sets again.

Reply:
Wooot!